### PR TITLE
Sanitize certificates from routes in the router

### PR DESCRIPTION
**Pull request by smarterclayton**
Ensure that no invalid PEM data is left in the route after extended
validation passes.

@liggitt

### CODE REVIEW

**Comment by smarterclayton** [comment]
@Miciah

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
@deads2k @sttts potential flake from rebase

https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin/1010/consoleFull#-108273047856bf4006e4b05b79524e5923

**Comment by smarterclayton** [comment]
any comments?  [test]

**Comment by deads2k** [comment]
I'm not familiar with the area, but I think @liggitt.  Let's give him today.

File: pkg/route/api/validation/validation.go [hunk 1]
```diff
@@ -111,8 +113,26 @@ func ValidateRouteStatusUpdate(route *routeapi.Route, older *routeapi.Route) fie
...
+func sanitizePEM(data []byte) []byte {
+	var block *pem.Block
+	buf := &bytes.Buffer{}
+	for len(data) > 0 {
+		block, data = pem.Decode(data)
+		if block == nil {
+			return buf.Bytes()
+		}
+		// ignores blocks that can't be serialized
+		_ = pem.Encode(buf, block)
```
Comment (liggitt) [inline pkg/route/api/validation/validation.go]: Open questions:

should we whitelist the PEM block types we allow ("CERTIFICATE", "PRIVATE KEY", "RSA PRIVATE KEY", etc)

When we're validating, are we making sure the data matches the PEM block types? Go's parsing sometimes ignores exact PEM block type matches and will parse any private key from any block type ending in "PRIVATE KEY"

which types does haproxy support?

Is haproxy forgiving about mismatched block type and block content?

**Comment by liggitt** [comment]
Sanitization is a strict improvement, but the questions about haproxy type support and type/content mismatch are important to find out before we open up certs

**Comment by smarterclayton** [comment]
Haproxy just calls down to OpenSSL.  So any block type supported which
is likely a huge list.

The mismatch between types is concerning - I can't think of a bounded
way to constrain that.

**Comment by smarterclayton** [comment]
Updated

File: pkg/route/api/validation/validation.go [hunk 2]
```diff
@@ -111,8 +115,94 @@ func ValidateRouteStatusUpdate(route *routeapi.Route, older *routeapi.Route) fie
 	return allErrs
 }
 
+type blockVerifierFunc func(block *pem.Block) error
+
+func publicKeyBlockVerifier(block *pem.Block) error {
```
Comment (liggitt) [inline pkg/route/api/validation/validation.go]: modifying in-place is a little unexpected... I'd expect a new block returned. there are other bits (block headers, etc) that this isn't clearing

File: pkg/route/api/validation/validation.go [hunk 3]
```diff
@@ -111,8 +115,94 @@ func ValidateRouteStatusUpdate(route *routeapi.Route, older *routeapi.Route) fie
...
+// to prevent OpenSSL or other non Golang tools from receiving unsanitized input.
+func sanitizePEM(data []byte) ([]byte, error) {
+	var block *pem.Block
+	buf := &bytes.Buffer{}
+	for len(data) > 0 {
+		block, data = pem.Decode(data)
+		if block == nil {
+			return buf.Bytes(), nil
+		}
+		fn, ok := knownBlockDecoders[block.Type]
```
Comment (liggitt) [inline pkg/route/api/validation/validation.go]: some ec private keys will include an `EC PARAMETERS` block that should be dropped

Comment (liggitt) [inline pkg/route/api/validation/validation.go]: c.f. https://github.com/kubernetes/kubernetes/issues/40589

Comment (liggitt) [inline pkg/route/api/validation/validation.go]: I'm wondering if we should just drop unrecognized block types and continue (as long as we got at least one recognized block)

**Comment by liggitt** [comment]
tests are a little light, a few nits, lgtm otherwise

File: pkg/route/api/validation/validation.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/route/api/validation/validation.go]: should it clear them?

Comment (liggitt) [inline pkg/route/api/validation/validation.go]: ummm... not sure? the uses I've seen for PEM headers in these types of blocks are for encrypted private keys, which we don't support

**Comment by smarterclayton** [comment]
more tests added.

**Comment by smarterclayton** [comment]
nits addressed

**Comment by liggitt** [comment]
LGTM

**Comment by smarterclayton** [comment]
[merge]

**Comment by smarterclayton** [comment]
[merge]

**Comment by smarterclayton** [comment]
[merge] install failure

**Pull request merged**